main/src/main/java/cgeo/geocaching/CacheDetailActivity.java

```diff
@@ -1847,6 +1853,101 @@ public void setContent() {
 
         }
 
+        MlKitTranslateUtil.LanguagePackDownloadHandler languagePackDownloadHandler = null;
+
+        private void initializeListingTranslator() {
+            if (Settings.getTranslationTargetLanguage().isEmpty()) {
+                binding.descriptionTranslate.setVisibility(View.GONE);
+                return;
+            }
+
+            final CacheDetailActivity cda = (CacheDetailActivity) getActivity();
+
+            // add observer to listing language
+            cda.listingTranslationLanguage.observe(this, lng -> {
+                if (null == lng || Settings.getLanguagesToNotTranslate().contains(lng.getCode())) {
+                    binding.descriptionTranslate.setVisibility(View.GONE);
+                } else if ("und".equals(lng.getCode())) {
+                    binding.descriptionTranslateButton.setEnabled(false);
+                    binding.descriptionTranslateNote.setText(R.string.translator_language_unknown);
+                } else {
+                    binding.descriptionTranslateButton.setEnabled(true);
+                    binding.descriptionTranslateNote.setText(getResources().getString(R.string.translator_language_detected, lng.toString()));
+                }
+            });
+            binding.descriptionTranslateButton.setOnClickListener(v -> translateListing());
+            binding.descriptionTranslateNote.setOnClickListener(v -> MlKitTranslateUtil.showLanguageSelection(getActivity(), cda.listingTranslationLanguage::setValue));
+            binding.descriptionTranslate.setVisibility(View.VISIBLE);
+
+            // identify listing language
+            LanguageIdentification.getClient().identifyLanguage(binding.description.getText().toString())
+                    .addOnSuccessListener(lngCode -> {
+                        cda.listingTranslationLanguage.setValue(new MlKitTranslateUtil.Language(lngCode));
+                    })
+                    .addOnFailureListener(e -> {
+                        binding.descriptionTranslateButton.setEnabled(false);
+                        binding.descriptionTranslateNote.setText(e.getMessage());
+                    });
+
+        }
+
+        private void translateListing() {
```

Comment:
I believe this method as a whole could be moved to the utils class. 
Eg a method "translate(Context context, CharSequence sourceText, Language sourceLanguage, BiConsumer<Language, String> translatedTextConsumer>) which transparent to the caller also handles all the "download stuff" logic (asking user if he wants to download etc etc)

Reply:
Everything that's left in this class is very specific to the UI of the specific activity. Out of 55 lines only 2 are calls to some translation stuff - and those 2 are the wrappers.

> which transparent to the caller also handles all the "download stuff" logic (asking user if he wants to download etc etc)

all of that is already in the utility method